File: crates/yryvu-bridge/src/commands/commits.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::path::PathBuf;

use graph_core::{build_pinned_set, layout_commits, Commit, GraphRow};
use serde::Serialize;
use tauri::{AppHandle, Emitter};

use crate::backend::{
    ApplyPatchOutcome, AuthorInfo, CombinedDiffSummary, CommitDetail, CommitDiff, GitBackend,
    ResetMode, RECENT_AUTHORS_DEFAULT_LIMIT,
};
use crate::repo::commits::{
    combined_commit_diff_summary as combined_commit_diff_summary_impl,
    commit_details as commit_details_impl, pick_pinned_head_for_path,
};
use crate::repo::hosting::{detect_hosting_service, parse_repo_identifiers};
use crate::repo::GixBackend;

/// Streamed batch payload for `graph:batch` events.
///
/// `pinned_sha` carries the trunk pin selected by `pick_pinned_head_for_path`
/// so the frontend can mark the pinned ref group with a pin annotation
/// (issue #145, doc 06 stage-2 ordering). Same value across every batch in
/// a single stream — frontend reads it from the first batch and ignores
/// repeats. `None` only when the repo has no resolvable trunk (detached HEAD
/// + no `origin/HEAD` + no canonical-name local branch).
#[derive(Debug, Clone, Serialize)]
pub struct GraphBatch {
    pub rows: Vec<GraphRow>,
    pub done: bool,
    #[serde(rename = "pinnedSha")]
    pub pinned_sha: Option<String>,
}

pub const GRAPH_BATCH_EVENT: &str = "graph:batch";

#[tauri::command]
pub async fn stream_graph(
    app: AppHandle,
    repo_path: String,
    batch_size: Option<usize>,
    limit: Option<usize>,
) -> Result<(), String> {
    let batch_size = batch_size.unwrap_or(100).max(1);
    let path = PathBuf::from(&repo_path);

    tauri::async_runtime::spawn_blocking(move || -> Result<(), String> {
        let backend = GixBackend;
        let walk = backend
            .walk_commits(&path, limit)
            .map_err(|e| e.to_string())?;
        let commits: Vec<Commit> = walk
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| e.to_string())?;

        let pinned_tip = pick_pinned_head_for_path(&path);
        let pinned_shas = build_pinned_set(&commits, pinned_tip.as_deref());

        let rows = layout_commits(commits, 32, pinned_shas).map_err(|e| e.to_string())?;

        let mut buffer: Vec<GraphRow> = Vec::with_capacity(batch_size);
        for row in rows {
            buffer.push(row);
            if buffer.len() >= batch_size {
                flush(&app, &mut buffer, false, pinned_tip.as_deref())
                    .map_err(|e| e.to_string())?;
            }
        }
        flush(&app, &mut buffer, true, pinned_tip.as_deref()).map_err(|e| e.to_string())?;
        Ok(())
    })
    .await
    .map_err(|e| e.to_string())??;

    Ok(())
}

fn flush(
    app: &AppHandle,
    buffer: &mut Vec<GraphRow>,
    done: bool,
    pinned_sha: Option<&str>,
) -> tauri::Result<()> {
    let rows = std::mem::take(buffer);
    app.emit(
        GRAPH_BATCH_EVENT,
        GraphBatch {
            rows,
            done,
            pinned_sha: pinned_sha.map(|s| s.to_string()),
        },
    )
}

#[tauri::command]
pub async fn commit_diff(repo_path: String, sha: String) -> Result<CommitDiff, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .commit_diff(&PathBuf::from(&repo_path), &sha)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

/// Multi-revision / WIP-aware diff for the right-panel inspector. Drives the
/// chip stats, file list, and header copy in one round-trip — see
/// [`crate::backend::CombinedDiff`] for the shape.
/// Metadata-only combined diff for the inspector's file list + stat chips
/// (#178). Same selection semantics as [`combined_commit_diff`] but the
/// rows carry no hunk bodies, so a huge WIP diff no longer serialises
/// every line across the IPC boundary for a view that never reads them.
/// The diff view fetches a file's hunks on demand (`commit_diff`) when the
/// user opens it.
#[tauri::command]
pub async fn combined_commit_diff_summary(
    repo_path: String,
    shas: Vec<String>,
    include_workdir: bool,
) -> Result<CombinedDiffSummary, String> {
    tauri::async_runtime::spawn_blocking(move || {
        combined_commit_diff_summary_impl(&PathBuf::from(&repo_path), &shas, include_workdir)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

/// Resolve full commit metadata for the right-panel inspector (message, body,
/// author, committer, timestamps, initials, gravatar hashes). Uses the gix
/// backend directly — no trait indirection since git2 fallback is not needed
/// for read-only metadata lookup.
#[tauri::command]
pub async fn commit_details(repo_path: String, sha: String) -> Result<CommitDetail, String> {
    tauri::async_runtime::spawn_blocking(move || {
        commit_details_impl(&PathBuf::from(&repo_path), &sha).map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn checkout_commit(repo_path: String, sha: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .checkout_commit(&PathBuf::from(&repo_path), &sha)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn reset_to_commit(
    repo_path: String,
    sha: String,
    mode: ResetMode,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .reset_to_commit(&PathBuf::from(&repo_path), &sha, mode)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn cherry_pick_commit(repo_path: String, sha: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .cherry_pick_commit(&PathBuf::from(&repo_path), &sha)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn cherry_pick_commits(
    repo_path: String,
    shas: Vec<String>,
    target_branch: Option<String>,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        let refs: Vec<&str> = shas.iter().map(String::as_str).collect();
        GixBackend
            .cherry_pick_commits_onto(&PathBuf::from(&repo_path), &refs, target_branch.as_deref())
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn revert_commit(repo_path: String, sha: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .revert_commit(&PathBuf::from(&repo_path), &sha)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn format_patch(
    repo_path: String,
    sha: String,
    out_dir: String,
) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .format_patch(&PathBuf::from(&repo_path), &sha, &PathBuf::from(&out_dir))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

/// Apply an mbox `.patch` (`git am` equivalent) onto HEAD as a new commit.
/// Needs `AppHandle` — unlike `format_patch` it creates a commit, so the
/// committer is stamped with the active profile's identity (author comes
/// from the mbox headers).
#[tauri::command]
pub async fn apply_patch(
    app: AppHandle,
    repo_path: String,
    patch_path: String,
) -> Result<ApplyPatchOutcome, String> {
    let config_dir = super::profiles::config_dir(&app).ok();
    tauri::async_runtime::spawn_blocking(move || {
        let path = PathBuf::from(&repo_path);
        let committer = config_dir
            .as_deref()
            .and_then(|dir| super::profiles::resolve_identity(dir, &path));
        let committer_ref = committer.as_ref().map(|(n, e)| (n.as_str(), e.as_str()));
        GixBackend
            .apply_patch(&path, &PathBuf::from(&patch_path), committer_ref)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

/// Classify the repo's primary remote into a hosting-service tag used by
/// the frontend to pick a provider-native avatar source (GitHub's CDN
/// email→avatar endpoint, GitLab's user API, Gravatar fallback, …).
///
/// Returns one of `"github" | "gitlab" | "bitbucket" | "gitea" | "unknown"`.
/// Never fails — bare repos / zero-remote repos / opaque self-hosted URLs
/// all collapse to `"unknown"` and the frontend falls back to Gravatar.
#[tauri::command]
pub async fn get_hosting_service(repo_path: String) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let path = PathBuf::from(&repo_path);
        Ok::<_, String>(detect_hosting_service(&path).as_str().to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

/// Provider classification + `(owner, repo)` for the active repo's
/// `origin` remote. Lets the frontend decide whether the PR /
/// Issues panels can fetch live data and on which provider.
///
/// `service` is always one of `"github" | "gitlab" | "bitbucket" |
/// "gitea" | "unknown"`. `owner` / `repo` are `None` for bare repos,
/// zero-remote repos, or URLs that don't split cleanly into two path
/// segments.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RepoProviderInfo {
    pub service: String,
    pub owner: Option<String>,
    pub repo: Option<String>,
}

#[tauri::command]
pub async fn get_repo_provider_info(repo_path: String) -> Result<RepoProviderInfo, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let path = PathBuf::from(&repo_path);
        let service = detect_hosting_service(&path).as_str().to_string();
        let (owner, repo) = gix::open(&path)
            .ok()
            .and_then(|r| crate::repo::hosting::remote_url(&r, "origin"))
            .and_then(|url| parse_repo_identifiers(&url))
            .map(|(o, r)| (Some(o), Some(r)))
            .unwrap_or((None, None));
        Ok::<_, String>(RepoProviderInfo {
            service,
            owner,
            repo,
        })
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn recent_authors(
    repo_path: String,
    limit: Option<usize>,
) -> Result<Vec<AuthorInfo>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .recent_authors(
                &PathBuf::from(&repo_path),
                limit.unwrap_or(RECENT_AUTHORS_DEFAULT_LIMIT),
            )
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

File: apps/yryvu-app/src/ipc/commits.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export function checkoutCommit(repoPath: string, sha: string): Promise<void> {
  return invoke<void>("checkout_commit", { repoPath, sha });
}

export type ResetMode = "soft" | "mixed" | "hard";

export function resetToCommit(
  repoPath: string,
  sha: string,
  mode: ResetMode,
): Promise<void> {
  return invoke<void>("reset_to_commit", { repoPath, sha, mode });
}

export function cherryPickCommit(repoPath: string, sha: string): Promise<void> {
  return invoke<void>("cherry_pick_commit", { repoPath, sha });
}

/**
 * Apply a batch of commits, optionally switching to `targetBranch` first.
 * `shas` are applied in array order (oldest → newest is the canonical
 * "in their original order" the issue #13 acceptance bullet asks for).
 * When `targetBranch` differs from the current HEAD label the working
 * tree must be clean — the backend returns the `WorkingTreeDirty`
 * variant and the UI is expected to gate / auto-stash as it does for
 * branch checkout. Mid-batch conflicts leave the repo on the target
 * branch with N-1 picks applied and CHERRY_PICK_HEAD on the failing
 * commit, so the StateBanner can surface the abort affordance.
 */
export function cherryPickCommits(
  repoPath: string,
  shas: string[],
  targetBranch: string | null,
): Promise<void> {
  return invoke<void>("cherry_pick_commits", {
    repoPath,
    shas,
    targetBranch,
  });
}

export function revertCommit(repoPath: string, sha: string): Promise<void> {
  return invoke<void>("revert_commit", { repoPath, sha });
}

export function formatPatch(
  repoPath: string,
  sha: string,
  outDir: string,
): Promise<string> {
  return invoke<string>("format_patch", { repoPath, sha, outDir });
}

/**
 * Provider tags recognised by the backend's `detect_hosting_service`.
 * Drives provider-native avatar resolution — notably the GitHub CDN
 * endpoint `avatars.githubusercontent.com/u/e?email=…` which resolves an
 * avatar from an email with no API auth (bundle offset 1508073 in
 * GitKraken's app bundle). Anything the backend can't confidently
 * classify collapses to `"unknown"` and the frontend falls back to
 * Gravatar.
 */
export type HostingService =
  | "github"
  | "gitlab"
  | "bitbucket"
  | "gitea"
  | "unknown";

export function getHostingService(repoPath: string): Promise<HostingService> {
  return invoke<HostingService>("get_hosting_service", { repoPath });
}

/**
 * Provider classification + `(owner, repo)` for the active repo's
 * `origin` remote. Used by the integration panels (PR list / Issues)
 * to know whether they can fetch live data and on which provider.
 *
 * `owner` / `repo` are `null` for bare repos, zero-remote repos, or
 * URLs whose path doesn't split cleanly into two segments.
 */
export interface RepoProviderInfo {
  service: HostingService;
  owner: string | null;
  repo: string | null;
}

export function getRepoProviderInfo(repoPath: string): Promise<RepoProviderInfo> {
  return invoke<RepoProviderInfo>("get_repo_provider_info", { repoPath });
}

export interface RefTag {
  name: string;
  kind: "Branch" | "RemoteBranch" | "Tag" | "Head";
  /**
   * Short name of the tracked remote branch (e.g. `origin/main`). `null` when
   * the local branch has no upstream configured or this ref isn't a local
   * branch. Drives the upstream indicator section of the composite ref pill
   * (issue #145, doc 06).
   */
  upstream: string | null;
  /**
   * Commits this local branch has that the upstream doesn't. `0` when no
   * upstream is configured. Renders as `↑N` in the composite pill.
   */
  ahead: number;
  /**
   * Commits the upstream has that this local branch doesn't. Renders as `↓N`.
   */
  behind: number;
}

/**
 * Mirror of the Rust `graph_core::ChildRefs` — refs reachable from this
 * commit's strict descendants, bucketed by kind. Populated via the
 * `populate_child_refs` pass in `layout_commits`. Rust-side `HashSet<String>`
 * serializes as a JSON array, so these land as `string[]` here.
 */
export interface ChildRefs {
  heads: string[];
  remotes: string[];
  tags: string[];
}

export interface GraphRow {
  sha: string;
  short_sha: string;
  summary: string;
  /**
   * Commit message body — everything after the subject line with leading
   * blank-line separators trimmed. Empty string when the commit has only a
   * subject. Raw content (trailers included) to match GitKraken's render,
   * which emojify-only transforms the body and parses trailers client-side
   * for the co-authors list.
   */
  body: string;
  /**
   * Author's display name only (no email). For the old `"Name <email>"` form
   * the backend used to send, compose via `${author_name} <${author_email}>`.
   */
  author_name: string;
  /** Author email, lowercase-trimmed by the backend via gix. */
  author_email: string;
  /**
   * Two-character initials fallback for the avatar badge, pre-computed in
   * `graph-core::author_initials`. Uppercased.
   */
  author_initials: string;
  /**
   * Lowercase hex MD5 of the trimmed-lowercased email — the Gravatar URL
   * fragment. Compose the full URL via
   * `https://gravatar.com/avatar/${gravatar_hash}?s=36&d=404`. `d=404` so
   * Gravatar returns 404 (instead of its default placeholder) when no
   * avatar is registered, letting the frontend fall back cleanly.
   */
  gravatar_hash: string;
  author_date: number;
  /**
   * Committer identity. Populated when the commit has committer metadata
   * (virtually always); `null` only on malformed / ancient commits the gix
   * backend couldn't decode. The right-panel commit block renders the
   * committer row only when both conditions hold: `committer_*` is non-null
   * AND differs from `author_name`/`author_email`. Mirrors the GitKraken
   * bundle guard `!committerInfo || (email===author && name===author)`.
   */
  committer_name: string | null;
  committer_email: string | null;
  committer_date: number | null;
  /**
   * Pre-computed initials + gravatar hash for the committer, same shape as
   * the author equivalents. `null` when committer metadata is absent.
   * Identical to the author values when committer matches author — frontend
   * still renders at most one block per the guard above.
   */
  committer_initials: string | null;
  committer_gravatar_hash: string | null;
  lane: number;
  parent_lanes: number[];
  parent_shas: string[];
  color_idx: number;
  refs: RefTag[];
  is_merge: boolean;
  child_refs: ChildRefs;
  /**
   * Lane indices carrying a visual edge through this row (sorted ascending,
   * deduplicated). Populated in graph-core via pre+post snapshots of
   * `columns_used` around each `place()` call. Consumed by the per-row
   * renderer (#81) to emit a vertical pipe segment per listed lane confined
   * to the row's height.
   */
  active_lanes: number[];
}

/**
 * Commit metadata powering the right-panel inspector (issue #112). Resolved
 * fresh from the repo via `commit_details` — not reused from a cached
 * `GraphRow` — so the inspector still works for commits outside the
 * current stream window.
 */
export interface CommitDetail {
  sha: string;
  /**
   * 6-character SHA prefix, matching GitKraken's inspector. `GraphRow.short_sha`
   * still uses 7 chars for graph-row consumers.
   */
  short_sha: string;
  parent_shas: string[];
  summary: string;
  body: string;
  author_name: string;
  author_email: string;
  author_date: number;
  author_initials: string;
  gravatar_hash: string;
  committer_name: string | null;
  committer_email: string | null;
  committer_date: number | null;
  committer_initials: string | null;
  committer_gravatar_hash: string | null;
}

export function getCommitDetails(
  repoPath: string,
  sha: string,
): Promise<CommitDetail> {
  return invoke<CommitDetail>("commit_details", { repoPath, sha });
}

interface GraphBatch {
  rows: GraphRow[];
  done: boolean;
  /**
   * SHA of the trunk pin selected by the backend's `pick_pinned_head`. Same
   * value across every batch in a stream — frontend writes it once on the
   * first batch that carries it. `null` when the repo has no resolvable
   * trunk (detached HEAD + no remote default + no canonical branch).
   */
  pinnedSha: string | null;
}

export interface StreamHandle {
  stop: () => void;
  promise: Promise<void>;
}

/**
 * Invokes `stream_graph` on the Rust side and emits row batches via `onBatch`.
 * Each batch also carries the trunk-pin sha through `onPinned`; identical
 * across batches for a given stream so the consumer can write it once and
 * ignore repeats. Resolves when the Rust task signals `done`.
 */
export function streamGraph(
  repoPath: string,
  onBatch: (rows: GraphRow[]) => void,
  options: {
    batchSize?: number;
    onPinned?: (sha: string | null) => void;
  } = {},
): StreamHandle {
  const batchSize = options.batchSize ?? 100;
  const onPinned = options.onPinned;
  let unlisten: UnlistenFn | undefined;
  let resolve!: () => void;
  let reject!: (err: unknown) => void;

  const promise = new Promise<void>((res, rej) => {
    resolve = res;
    reject = rej;
  });

  (async () => {
    try {
      unlisten = await listen<GraphBatch>("graph:batch", (e) => {
        onPinned?.(e.payload.pinnedSha);
        if (e.payload.rows.length > 0) onBatch(e.payload.rows);
        if (e.payload.done) {
          unlisten?.();
          resolve();
        }
      });
      await invoke("stream_graph", { repoPath, batchSize });
    } catch (err) {
      unlisten?.();
      reject(err);
    }
  })();

  return {
    stop: () => unlisten?.(),
    promise,
  };
}

/// Mirrors `yryvu_bridge::backend::AuthorInfo`. One entry per unique
/// author (deduped by lowercased email) found in the last N commits.
export interface AuthorInfo {
  name: string;
  email: string;
  count: number;
}

/// Walks recent commits and returns unique authors ordered by
/// frequency. Drives the commit panel's "Add Co-Author" picker. Pass
/// `undefined` for `limit` to use the backend default (200 commits).
export function recentAuthors(
  repoPath: string,
  limit?: number
): Promise<AuthorInfo[]> {
  return invoke<AuthorInfo[]>("recent_authors", { repoPath, limit });
}
